lib/HTML/Restrict.pm

```diff
@@ -150,16 +150,34 @@ sub _build_parser {
 
                     foreach my $source_type ( 'href', 'src', 'cite' ) {
 
-                        if ( $attr->{$source_type} ) {
-                            my $uri = URI->new( $attr->{$source_type} );
-                            if ( defined $uri->scheme ) {
+                        my $link = $attr->{$source_type};
+
+                        # Remove unprintable ASCII control characters, which
+                        # are 1..8 and 14..31. These characters are not valid
+                        # in URLs, but they can prevent the URI parser from
+                        # recognizing the scheme when they are used as leading
+                        # characters.  Browsers will helpfully ignore them,
+                        # meaning that these characters can be used to defeat
+                        # HTML::Restrict when used as leading characters in a
+                        # link.  In our case we will strip them all regardless
+                        # of where they are in the URL. See
+                        # https://github.com/oalders/html-restrict/issues/30
+
+                        if ($link) {
+                            $link =~ s/[\001-\010]/ /g;    # decimal 1..8
+                            $link =~ s/[\016-\037]/ /g;    # decimal 14..31
```

Comment:
Is there a reason why you picked this set of characters? Does the null byte or a vertical tab not have the same issues? What about Unicode control characters? I wonder if something like `[[:cntrl:]]` or `[^[:print:]]` would be preferable.

Reply:
So, the null byte doesn't suffer from this problem.  (At least not in Chrome). Assuming `<a href="&#0;javascript:alert(1);">0</a><br>` is the correct way to test it.

However, I did some more reading at https://infra.spec.whatwg.org/#c0-control and it looks like "U+0000 NULL to U+001F INFORMATION SEPARATOR ONE" inclusive are not valid in a URL, so there's no reason not to include the whole set.  I'll make some changes.

There is no mention of C1 control codes at the spec I've been working from: https://url.spec.whatwg.org/

Reply:
It doesn't work for the null byte, but it does work for the vertical tab and other control characters that are currently being excluded.

Reply:
Interesting.  I tested them on Mac Chrome, Firefox and Safari and none of the excluded characters created a dialog box.

Reply:
I've reworked the PR to remove the whole set.

Reply:
Weird. I tested`<a href="&#0011;javascript:alert(1);">0</a>` and `<a href="&#xB;javascript:alert(1);">0</a>` on both Firefox and Chrome on Ubuntu and saw the dialog with both.